File: backend/pkg/gameplugin/sdk/go/event.go
package sdk

import (
	"context"
	"encoding/json"

	"github.com/u-ai/backend/pkg/gameplugin/protocol"
)

const MethodEventPublish = "plugin.event.publish"
const PluginAgentEventID = "plugin.agent_event"

type EventPublishInput struct {
	ChannelID string                     `json:"channelId"`
	EventID   string                     `json:"eventId"`
	Payload   json.RawMessage            `json:"payload"`
	Metadata  map[string]json.RawMessage `json:"metadata,omitempty"`
}

func (c *Client) PublishEvent(ctx context.Context, input EventPublishInput, opts ...MessageOption) (protocol.Envelope, error) {
	payload := map[string]any{"channelId": input.ChannelID, "eventId": input.EventID, "payload": input.Payload}
	if input.Metadata != nil {
		payload["metadata"] = input.Metadata
	}
	return c.sendHostNotification(ctx, MethodEventPublish, payload, opts...)
}

// PublishAgentEvent publishes a generic plugin-defined event that may wake the
// bound Agent context. GameHost treats Type and Payload as opaque data.
func (c *Client) PublishAgentEvent(ctx context.Context, event protocol.PluginEvent, metadata map[string]json.RawMessage, opts ...MessageOption) (protocol.Envelope, error) {
	payload, err := json.Marshal(event)
	if err != nil {
		return protocol.Envelope{}, err
	}
	return c.PublishEvent(ctx, EventPublishInput{ChannelID: event.SessionID, EventID: PluginAgentEventID, Payload: payload, Metadata: metadata}, opts...)
}

File: backend/pkg/gameplugin/sdk/game-plugin/src/event.ts
import { Envelope } from './protocol';
import { Client, MessageOption } from './client';
import type { PluginEvent } from './game';

export const METHOD_EVENT_PUBLISH = 'plugin.event.publish';
export const PLUGIN_AGENT_EVENT_ID = 'plugin.agent_event';

export interface EventPublishInput {
  channelId: string;
  eventId: string;
  payload: unknown;
  metadata?: Record<string, unknown>;
}

export async function publishEvent(
  client: Client,
  input: EventPublishInput,
  opts: MessageOption[] = []
): Promise<Envelope> {
  const payload: Record<string, unknown> = {
    channelId: input.channelId,
    eventId: input.eventId,
    payload: input.payload,
  };
  if (input.metadata) {
    payload.metadata = input.metadata;
  }
  return client.sendReservedNotification(METHOD_EVENT_PUBLISH, payload, ...opts);
}

export async function publishAgentEvent(
  client: Client,
  event: PluginEvent,
  metadata: Record<string, unknown> = {},
  opts: MessageOption[] = []
): Promise<Envelope> {
  return publishEvent(client, {
    channelId: event.sessionId ?? '',
    eventId: PLUGIN_AGENT_EVENT_ID,
    payload: event,
    metadata,
  }, opts);
}
